text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+-   `Send` and `Sync` should definitely be kept out. `Allocator`-based stores could not use `NonNull<u8>` otherwise.
+
+
+##  (Minor) Should `Store::dangling` be `const`?
+
+While const trait associated functions are still a maybe, it seems reasonable to ask ourselves whether some of the
+associated functions of `Store` should be `const` if it were possible.
+
+There doesn't seem to be a practical advantage in doing so for most of the associated functions of `Store`: if
+allocation and deallocation need be executed in a const context, then a `const Store` is necessary, and there's no need
+to single out any of those.
+
+There is, however, a very practical advantage in making `Store::dangling` const: it allows initializing an empty
+collection in a const context even with a non-const `Store` implementation.
+
+The one downside is that this would preclude some implementations of `dangling` which would rely on global state, or
+I/O. @CAD97 notably mentioned the possibility of using randomization for debugging or hardening purposes. Still, it
+would still be possible to initialize the instance of `Store` with a random seed, then use a PRNG within `dangling`.
+
+
+#   Future Possibilities
+
+##  SharingStore
+
+One (other) underdevelopped aspect of the `Allocator` API at the moment is the handling of fungibility of pointers, that
+is the description -- in trait -- of whether a pointer allocated by one `Allocator` can be grown, shrunk, or deallocated
+by another instance of `Allocator`. The immediate consequence is that `Rc` is only `Clone` for `Global`, and the
+`LinkedList::append` method is similarly only available for `Global` allocator.
+
+A possible future extension for the Storage proposal is the introduction of the `SharingStore` trait:
+
+```rust
+trait SharingStore: PinningStore {
+    type SharingError;
+
+    fn is_sharing_with(&self, other: &Self) -> bool;
+
+    fn share(&self) -> Result<Self, Self::SharingError> where Self: Sized;
+}
+```
+
+This trait introduces the concept of set of sharing stores, that is when multiple stores share the same "backing" memory
+and allocation metadata.
+
+The `share` method creates a new instance of the store which shares the same "backing" memory and metadata as `self`,
+while the `is_sharing_with` method allows querying whether two stores share the same "backing" memory and metadata.
+
+A set of sharing stores can be thought of as a single store instance: handles created by one of the stores can be used
+with any of the stores of the set, in any way, and as long as one store of the set has not been dropped, dropping a
+store of the set does not invalidate the handles. Informally, the "backing" memory and metadata can be thought of as
+being reference-counted.
+
+The requirement of `PinningStore` is necessary as moving any one instance should not invalidate the pointers resolved by
+other instances of the set, and the `SharingError` type allows modelling potentially-sharing stores, such as a small
```

Comment:
`SharingStore` doesn't strictly require the full `PinningStore`; the guarantee is somewhere between `StableStore` and `PinningStore`. Namely, moving clones mustn't invalidate resolved pointers, but `forget`ting all of the clones is fine if you have `SharingStore` but not `PinningStore`.

I might even argue that `StableStore` should be this, rather than still allowing moves of the storage to invalidate resolved pointers. This is a guarantee I can see the utility of, unlike the `StableStore` which doesn't provide stability across moves of the `Store` object.

Reply:
I agree w.r.t. implementation. Where I disagree is consumers: what is actually, functionally gained by adding a requirement on `StableStore` but not `PinningStore`?

Reply:
Everything.

Let's take a `LinkedList`:

```rust
let list = /* somehow [0, 1, 2, 3, 4] */

let x0 = list.get(0);
let x1 = list.get(1);
```

With a non-`StableStore`, the second dereference -- and thus second call to `resolve` with a different handle -- invalidates the first reference (stored in `x0`). Thus you can never have two references to distinct elements at once. This is _very_ limiting, needless to say. You can't even implement an `Iterator`.

Now, a `PinningStore` doesn't have the issue, but you can't have an in-line store be a `PinningStore` because moving the in-line store moves all the elements around, invalidating former pointers.

And thus `StableStore` was born. An in-line store can be a `StableStore`, and can therefore be used for `LinkedList`, `BTreeSet`, `BTreeMap`, `HashSet`, `HashMap`, etc... and get nigh-fully functional APIs.

Reply:
I think I disagree that `LinkedList` wants to use a single `StableStore + MultipleStore`; the key reason to use a linked list is that a pointer to an entry in the list alone is sufficient to manipulate the list (add/remove/splice/split), i.e. without also a pointer to a shared allocator handle.

I'll agree that collections managing multiple allocations want to be able to call `resolve` multiple times without invalidating previously resolved pointers.

Part of the reason for the slight disagreement in specifics is probably that v2 (and my partial v4 draft), in having separate `&self` and `&mut self` resolves, with only the latter invalidating resolved pointers and the former never doing so, meaning repeated shared access is permitted even without `StableStore`; only for separate-resolve split unique-tagged access does it require `StableStore`. And at the collection level, you have `&mut` for `DerefMut` access anyway.

Getting away without `StableStore` is barely achievable by the fact that `resolve`ing the same handle doesn't invalidate previously resolved pointers to that same handle. Which also places a notable limit on implementations, in that while `resolve` may result in other handles being relocated, it mustn't relocate the handle being resolved.

Reply:
> Getting away without StableStore is barely achievable by the fact that resolveing the same handle doesn't invalidate previously resolved pointers to that same handle. Which also places a notable limit on implementations, in that while resolve may result in other handles being relocated, it mustn't relocate the handle being resolved.

Indeed, resolving the same handle in a row -- with no intervening method call, nor move -- yielding the same pointer is the one memory-stability guarantee I baked in the base trait.

I do think, still, that `StableStore` is necessary to allow element-access (and thus `Iterator`) to be implemented on collections, which is fairly basic operation.

Reply:
This is the case for `MultipleStorage`. The use cases for {`StableStore` without `MultipleStore`} or {`MultipleStore` without `StableStore`} continue to elude me, but I'll acquiesce that to my breadth of collection-ish types implemented only covering those which want to always use a single allocation or want each allocation to have its own allocation handle.

Reply:
The `MultipleStore` without `StableStore` is sufficient to implement a `Cursor`-like API.

Unlike an `Iterator` API, a `Cursor`-like API only allows accessing one element at a time -- since a cursor is borrowed for the lifetime of any reference it returns, and must be mutated to point to another element -- and therefore cares not for the guarantees that `StableStore` offers.

Databases typically only offer `Cursor`-like APIs (or `Iterator` returning owned elements) to avoid blowing up the memory, for example, and thus you could implement a `Store` over a database, with a LRU cache of recently accessed elements, and power with it an associative container with Cursor but no Iterator.

I am not sure that a `StableStore` without `MultipleStore` ever makes sense, however.

In the absence of `MultipleStore`, any handle -- and any pointer derived from that handle -- is invalidated when performing a call to any allocation/deallocation function, leaving only `dangling` and `resolve` as the non handle-invalidating methods... and `StableStore` adds no further guarantee for those 2 methods.

I would still advocate for NOT having `StableStore` require `MultipleStore` though, given that they provide guarantees along different axes (resp. pointer invalidation & handle invalidation).